include/fiction/algorithms/physical_design/post_layout_optimization.hpp

```diff
@@ -1035,6 +531,567 @@ class post_layout_optimization_impl
      * Statistics about the post-layout optimization process.
      */
     post_layout_optimization_stats& pst;
+    /**
+     * Maximum number of relocations to try for each gate.
+     */
+    std::optional<uint64_t> max_gate_relocations = std::nullopt;
+    /**
+     * Maximum coordinate of all gates that are not POs.
+     */
+    tile<Lyt> max_non_po{0, 0};
+    /**
+     * Timeout limit reached.
+     */
+    bool timeout_limit_reached = false;
+    /**
+     * Start time.
+     */
+    std::chrono::time_point<std::chrono::high_resolution_clock> start;
+    /**
+     * Utility function to move wires that cross over empty tiles down one layer. This can happen if the wiring of a
+     * gate is deleted.
+     *
+     * @param lyt Gate-level layout.
+     * @param deleted_coords Tiles that got deleted.
+     */
+    void fix_wires(ObstrLyt& lyt, const std::vector<tile<ObstrLyt>>& deleted_coords) noexcept
+    {
+        std::unordered_set<tile<ObstrLyt>> moved_tiles{};
```

Comment:
You could try to use `phmap::parallel_flat_hash_set` here instead of `std::unordered_set` to see if you get a performance improvement. Analogously for your other usages of associative STL containers.